# Aviation Accidents Analysis

You are part of a consulting firm that is tasked to do an analysis of commercial and passenger jet airline safety. The client (an airline/airplane insurer) is interested in knowing what types of aircraft (makes/models) exhibit low rates of total destruction and low likelihood of fatal or serious passenger injuries in the event of an accident. They are also interested in any general variables/conditions that might be at play. Your analysis will be based off of aviation accident data accumulated from the years 1948-2023. 

Our client is only interested in airplane makes/models that are professional builds and could potentially still be active. Assume a max lifetime of 40 years for a make/model retirement and make sure to filter your data accordingly (i.e. from 1983 onwards). They would also like separate recommendations for small aircraft vs. larger passenger models. **In addition, make sure that claims that you make are statistically robust and that you have enough samples when making comparisons between groups.**


In this summative assessment you will demonstrate your ability to:
- **Use Pandas to load, inspect, and clean the dataset appropriately.**
- **Transform relevant columns to create measures that address the problem at hand.**
- conduct EDA: visualization and statistical measures to systematically understand the structure of the data
- recommend a set of airplanes and makes conforming to the client's request and identify at least *two* factors contributing to airplane safety. You must provide supporting evidence (visuals, summary statistics, tables) for each claim you make.

### Make relevant library imports

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

## Data Loading and Inspection

### Load in data from the relevant directory and inspect the dataframe.
- inspect NaNs, datatypes, and summary statistics

In [2]:
#with open("data/AviationData.csv", 'rb') as f:
    

In [3]:
aviation_df = pd.read_csv("data/AviationData.csv", encoding="utf-8", encoding_errors="replace")
aviation_df.head()

/var/folders/th/xsvfrj3d3bvffr_vt4sx1f7m0000gn/T/ipykernel_7739/3007147420.py:1: DtypeWarning: Columns (6,7,28) have mixed types. Specify dtype option on import or set low_memory=False.
  aviation_df = pd.read_csv("data/AviationData.csv", encoding="utf-8", encoding_errors="replace")


,Event.Id,Investigation.Type,Accident.Number,Event.Date,Location,Country,Latitude,Longitude,Airport.Code,Airport.Name,...,Purpose.of.flight,Air.carrier,Total.Fatal.Injuries,Total.Serious.Injuries,Total.Minor.Injuries,Total.Uninjured,Weather.Condition,Broad.phase.of.flight,Report.Status,Publication.Date
0,20001218X45444,Accident,SEA87LA080,1948-10-24,"MOOSE CREEK, ID",United States,NaN,NaN,NaN,NaN,...,Personal,NaN,2.0,0.0,0.0,0.0,UNK,Cruise,Probable Cause,NaN
1,20001218X45447,Accident,LAX94LA336,1962-07-19,"BRIDGEPORT, CA",United States,NaN,NaN,NaN,NaN,...,Personal,NaN,4.0,0.0,0.0,0.0,UNK,Unknown,Probable Cause,19-09-1996
2,20061025X01555,Accident,NYC07LA005,1974-08-30,"Saltville, VA",United States,36.922223,-81.878056,NaN,NaN,...,Personal,NaN,3.0,NaN,NaN,NaN,IMC,Cruise,Probable Cause,26-02-2007
3,20001218X45448,Accident,LAX96LA321,1977-06-19,"EUREKA, CA",United States,NaN,NaN,NaN,NaN,...,Personal,NaN,2.0,0.0,0.0,0.0,IMC,Cruise,Probable Cause,12-09-2000
4,20041105X01764,Accident,CHI79FA064,1979-08-02,"Canton, OH",United States,NaN,NaN,NaN,NaN,...,Personal,NaN,1.0,2.0,NaN,0.0,VMC,Approach,Probable Cause,16-04-1980


In [4]:
#inspect NaNs, datatypes
aviation_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88889 entries, 0 to 88888
Data columns (total 31 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Event.Id                88889 non-null  object 
 1   Investigation.Type      88889 non-null  object 
 2   Accident.Number         88889 non-null  object 
 3   Event.Date              88889 non-null  object 
 4   Location                88837 non-null  object 
 5   Country                 88663 non-null  object 
 6   Latitude                34382 non-null  object 
 7   Longitude               34373 non-null  object 
 8   Airport.Code            50132 non-null  object 
 9   Airport.Name            52704 non-null  object 
 10  Injury.Severity         87889 non-null  object 
 11  Aircraft.damage         85695 non-null  object 
 12  Aircraft.Category       32287 non-null  object 
 13  Registration.Number     87507 non-null  object 
 14  Make                    88826 non-null

In [5]:
#summary statistics
aviation_df.describe()

,Number.of.Engines,Total.Fatal.Injuries,Total.Serious.Injuries,Total.Minor.Injuries,Total.Uninjured
count,82805.000000,77488.000000,76379.000000,76956.000000,82977.000000
mean,1.146585,0.647855,0.279881,0.357061,5.325440
std,0.446510,5.485960,1.544084,2.235625,27.913634
min,0.000000,0.000000,0.000000,0.000000,0.000000
25%,1.000000,0.000000,0.000000,0.000000,0.000000
50%,1.000000,0.000000,0.000000,0.000000,1.000000
75%,1.000000,0.000000,0.000000,0.000000,2.000000
max,8.000000,349.000000,161.000000,380.000000,699.000000


## Data Cleaning

### Filtering aircrafts and events

We want to filter the dataset to include aircraft that the client is interested in an analysis of:
- inspect relevant columns
- figure out any reasonable imputations
- filter the dataset

1. ##### Filter to include only relevant years
Our client is only interested in airplane makes/models that are professional builds and could potentially still be active. Assume a max lifetime of 40 years for a make/model retirement and make sure to filter your data accordingly (i.e., from 1983 onwards).

In [6]:
relevant_columns = [
    "Event.Id",
    "Event.Date",
    "Aircraft.Category",
    "Amateur.Built",
    "Make",
    "Model",
    "Aircraft.damage",
    "Injury.Severity",
    "Total.Fatal.Injuries",
    "Total.Serious.Injuries",
    "Total.Minor.Injuries",
    "Total.Uninjured",
    "Number.of.Engines",
    "Engine.Type",
    "FAR.Description",
    "Weather.Condition",
    "Broad.phase.of.flight",
    "Purpose.of.flight",
]

# Select only relevant columns
relevant_df = aviation_df[relevant_columns].copy()

In [7]:
relevant_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88889 entries, 0 to 88888
Data columns (total 18 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Event.Id                88889 non-null  object 
 1   Event.Date              88889 non-null  object 
 2   Aircraft.Category       32287 non-null  object 
 3   Amateur.Built           88787 non-null  object 
 4   Make                    88826 non-null  object 
 5   Model                   88797 non-null  object 
 6   Aircraft.damage         85695 non-null  object 
 7   Injury.Severity         87889 non-null  object 
 8   Total.Fatal.Injuries    77488 non-null  float64
 9   Total.Serious.Injuries  76379 non-null  float64
 10  Total.Minor.Injuries    76956 non-null  float64
 11  Total.Uninjured         82977 non-null  float64
 12  Number.of.Engines       82805 non-null  float64
 13  Engine.Type             81793 non-null  object 
 14  FAR.Description         32023 non-null

In [8]:
relevant_df.describe()

,Total.Fatal.Injuries,Total.Serious.Injuries,Total.Minor.Injuries,Total.Uninjured,Number.of.Engines
count,77488.000000,76379.000000,76956.000000,82977.000000,82805.000000
mean,0.647855,0.279881,0.357061,5.325440,1.146585
std,5.485960,1.544084,2.235625,27.913634,0.446510
min,0.000000,0.000000,0.000000,0.000000,0.000000
25%,0.000000,0.000000,0.000000,0.000000,1.000000
50%,0.000000,0.000000,0.000000,1.000000,1.000000
75%,0.000000,0.000000,0.000000,2.000000,1.000000
max,349.000000,161.000000,380.000000,699.000000,8.000000


In [9]:
#total nulls in the injury severity column
relevant_df['Injury.Severity'].isnull().sum()

1000

In [10]:
#fill nulls with 'UNKNOWN' and standardize values with .upper
relevant_df["Injury.Severity"] = (
    relevant_df["Injury.Severity"].fillna("UNKNOWN").astype(str).str.strip().str.upper()
)
relevant_df['Injury.Severity'].value_counts()

Injury.Severity
NON-FATAL     67357
FATAL(1)       6167
FATAL          5262
FATAL(2)       3711
INCIDENT       2219
              ...  
FATAL(80)         1
FATAL(217)        1
FATAL(169)        1
FATAL(88)         1
FATAL(189)        1
Name: count, Length: 110, dtype: int64

### Cleaning and constructing Key Measurables

Injuries and robustness to destruction are a key interest point for the client. Clean and impute relevant columns and then create derived fields that best quantifies what the client wishes to track. **Use commenting or markdown to explain any cleaning assumptions as well as any derived columns you create.**

**Construct metric for fatal/serious injuries**

*Hint:* Estimate the total number of passengers on each flight. The likelihood of serious / fatal injury can be estimated as a fraction from this.

In [11]:
#obtain relevant year using split and take the first index
relevant_df["Year"] = (
    relevant_df["Event.Date"].str.split("-").str[0].astype(int)
)


In [12]:
# Convert to datetime
relevant_df["Event.Date"] = pd.to_datetime(
    relevant_df["Event.Date"], errors="coerce"
)

# Filter for active years (1983+)
filtered_df = relevant_df[relevant_df["Event.Date"].dt.year >= 1983].copy()

In [13]:
injury_cols = [
    "Total.Fatal.Injuries",
    "Total.Serious.Injuries",
    "Total.Minor.Injuries",
    "Total.Uninjured",
]

#Assumption: Blank entries  represent 0 count for that category
filtered_df[injury_cols] = filtered_df[injury_cols].fillna(0)

# Estimate total onboard passengers and crew
filtered_df["Total_Occupants"] = filtered_df[injury_cols].sum(axis=1)

In [14]:
filtered_df.head()

,Event.Id,Event.Date,Aircraft.Category,Amateur.Built,Make,Model,Aircraft.damage,Injury.Severity,Total.Fatal.Injuries,Total.Serious.Injuries,Total.Minor.Injuries,Total.Uninjured,Number.of.Engines,Engine.Type,FAR.Description,Weather.Condition,Broad.phase.of.flight,Purpose.of.flight,Year,Total_Occupants
3600,20001214X42040,1983-01-01,NaN,No,Piccard,AX-6,NaN,NON-FATAL,0.0,1.0,0.0,1.0,NaN,Unknown,NaN,VMC,Landing,Personal,1983,2.0
3601,20001214X42095,1983-01-01,NaN,No,Cessna,182P,Substantial,NON-FATAL,0.0,0.0,1.0,3.0,1.0,Reciprocating,NaN,VMC,Approach,Personal,1983,4.0
3602,20001214X42067,1983-01-01,NaN,No,Cessna,182RG,Substantial,NON-FATAL,0.0,0.0,0.0,2.0,1.0,Reciprocating,NaN,VMC,Landing,Personal,1983,2.0
3603,20001214X42063,1983-01-01,NaN,No,Cessna,182P,Substantial,NON-FATAL,0.0,0.0,0.0,1.0,1.0,Reciprocating,NaN,VMC,Takeoff,Personal,1983,1.0
3604,20001214X42018,1983-01-01,NaN,No,Piper,PA-28R-200,Substantial,NON-FATAL,0.0,0.0,2.0,0.0,1.0,Reciprocating,NaN,VMC,Approach,Personal,1983,2.0


**Aircraft.Damage**
- identify and execute any cleaning tasks
- construct a derived column tracking whether an aircraft was destroyed or not.

In [15]:
#Clean values (fill missing with 'UNKNOWN', remove spaces, make uppercase)
filtered_df["Aircraft.damage"] = (
    filtered_df["Aircraft.damage"].fillna("UNKNOWN").str.strip().str.upper()
)


In [16]:
#View counts of different values in our column
filtered_df["Aircraft.damage"].value_counts()

Aircraft.damage
SUBSTANTIAL    61775
DESTROYED      17575
UNKNOWN         3257
MINOR           2682
Name: count, dtype: int64

#### construct a derived column tracking whether an aircraft was destroyed or not.
* **DESTROYED** = Hull Total Loss (Write-Off): The aircraft is beyond repair or completely consumed. The insurer pays out the full insured value (100% hull payout). This is the exact definition of "total destruction."

* **SUBSTANTIAL** = Hull Repair Claim: The aircraft suffered major structural damage (e.g., bent wing spars, crushed landing gear), but the airframe is repairable. The insurer pays for repairs/overhaul, which is usually a fraction of the total hull value.

I will isolate `Total Destruction` (Hull Loss) as events coded `DESTROYED` because these represent 100% total loss payouts for underwriters. `SUBSTANTIAL` damage events are analyzed separately as repair claims, ensuring our total loss risk model is not diluted by repairable airframe incidents

In [17]:
# 1. Create binary column (1 if DESTROYED, 0 for anything else)
filtered_df["Is_Destroyed"] = (
    filtered_df["Aircraft.damage"].astype(str).str.strip().str.upper() == "DESTROYED"
).astype(int)

# 2. Check value counts to confirm
print(filtered_df["Is_Destroyed"].value_counts())

Is_Destroyed
0    67714
1    17575
Name: count, dtype: int64


### Investigate the *Make* column
- Identify cleaning tasks here
- List cleaning tasks clearly in markdown
- Execute the cleaning tasks
- For your analysis, keep Makes with a reasonable number (you can put the threshold at 50 though lower could work as well)

### Data Cleaning Tasks: Professional & Active Aircraft Filter

| Cleaning Task | Operations Required | Rationale & Client Constraint |
| :--- | :--- | :--- |
| **1. Filter Professional Builds** | Filter `Amateur.Built == 'NO'` | Restricts scope to professionally manufactured aircraft, removing experimental kit builds. |
| **2. Standardize Manufacturer Names (`Make`)** | Trim whitespace, uppercase, and apply regex mapping | Consolidates spelling variations (e.g., `CESSNA AIRCRAFT CO`, `CESSNA CO`, `CESSNA`) into single canonical make names. |
| **3. Apply Statistical Threshold ($N \ge 50$)** | Filter `Make` counts where total incident volume $N \ge 50$ | Removes low-frequency manufacturers to prevent small-sample noise from skewing risk rates. |

#### Step 1: Filter to remove non professional builds
 * Our client is only interested in airplane makes/models that are professional builds and could potentially still be active.
 * An amateur-built aircraft—often called a homebuilt aircraft—is a plane that is constructed by an individual or group of individuals rather than built in a professional factory.
 * Lets explore `Amateur.Build column`

In [18]:
filtered_df['Amateur.Built'].value_counts()

Amateur.Built
No     76960
Yes     8228
Name: count, dtype: int64

In [19]:
#fill nulls with unknown
filtered_df['Amateur.Built'] = filtered_df['Amateur.Built'].fillna("Unknown")

In [20]:
# filtered_df['Amateur.Built'].value_counts() -> we have only 101 unknowns, its safe to drop them
#we take only professional builds represented by NO
filtered_df = filtered_df[
    filtered_df["Amateur.Built"].astype(str).str.strip().str.upper() == "NO"
]

#### Step2: Standardize name of 'Make'


In [21]:
#standardize make names 
filtered_df["Make"] = (
    filtered_df["Make"].fillna("UNKNOWN").astype(str).str.strip().str.upper()
)
filtered_df["Make"]

3600                        PICCARD
3601                         CESSNA
3602                         CESSNA
3603                         CESSNA
3604                          PIPER
                    ...            
88884                         PIPER
88885                      BELLANCA
88886    AMERICAN CHAMPION AIRCRAFT
88887                        CESSNA
88888                         PIPER
Name: Make, Length: 76960, dtype: object

In [22]:
# Mapping dictionary for duplicate/inconsistent manufacturer names
make_mapping = {
    # Airbus
    "AIRBUS INDUSTRIE": "AIRBUS",
    # Bombardier
    "BOMBARDIER INC": "BOMBARDIER",
    # Cirrus
    "CIRRUS DESIGN CORP.": "CIRRUS",
    "CIRRUS DESIGN CORP": "CIRRUS",
    # Robinson
    "ROBINSON HELICOPTER": "ROBINSON",
    "ROBINSON HELICOPTER COMPANY": "ROBINSON",
    # De Havilland
    "DEHAVILLAND": "DE HAVILLAND",
    # Grumman variants
    "GRUMMAN AMERICAN": "GRUMMAN",
    "GRUMMAN AMERICAN AVN. CORP.": "GRUMMAN",
    "GRUMMAN-SCHWEIZER": "GRUMMAN",
    "GRUMMAN ACFT ENG COR-SCHWEIZER": "GRUMMAN",
    # Ercoupe
    "ERCOUPE (ENG & RESEARCH CORP.)": "ERCOUPE",
    # Air Tractor
    "AIR TRACTOR INC": "AIR TRACTOR",
    # Aviat
    "AVIAT AIRCRAFT INC": "AVIAT",
    # Diamond
    "DIAMOND AIRCRAFT IND INC": "DIAMOND",
    # Rockwell
    "ROCKWELL INTERNATIONAL": "ROCKWELL",
    # Learjet
    "GATES LEARJET": "LEARJET",
    # Aerostar
    "SMITH, TED AEROSTAR": "AEROSTAR",
    # Flight Design
    "FLIGHT DESIGN GMBH": "FLIGHT DESIGN",
    # American Champion
    "AMERICAN CHAMPION AIRCRAFT": "AMERICAN CHAMPION",
    #BOEING
    "BOEING STEARMAN": "BOEING",
    #CHAMPION
    "AMERICAN CHAMPION": "CHAMPION",
}

# 1. Strip leading/trailing whitespace and convert to uppercase
filtered_df["Make"] = filtered_df["Make"].astype(str).str.strip().str.upper()

# 2. Replace variants with standard names
filtered_df["Make"] = filtered_df["Make"].replace(make_mapping)

#### Step4:Filter Make counts where total incident volume is >= 50


In [23]:
# 1. Count incidents per manufacturer
make_counts = filtered_df["Make"].value_counts()

# 2. Retain only Makes with at least 50 incidents
robust_makes = make_counts[make_counts >= 50].index
filtered_df = filtered_df[filtered_df["Make"].isin(robust_makes)].copy()

# 3. Check remaining rows and qualifying Makes count
print(f"Remaining records: {len(filtered_df):,}")
print(f"Qualifying Makes: {filtered_df['Make'].nunique()}")
print(filtered_df["Make"].value_counts().head(10))

Remaining records: 70,567
Qualifying Makes: 72
Make
CESSNA         25740
PIPER          14099
BEECH           5094
BOEING          2699
BELL            2594
ROBINSON        1621
GRUMMAN         1510
MOONEY          1271
BELLANCA         978
AIR TRACTOR      892
Name: count, dtype: int64


### Inspect Model column
- Get rid of any NaNs.
- Inspect the column and counts for each model/make. Are model labels unique to each make?
- If not, create a derived column that is a unique identifier for a given plane type.

In [24]:
#Get rid of any NaNs.
filtered_df = filtered_df.dropna(subset=['Model'])

In [25]:
# Group by Make and Model together
make_model_df = (
    filtered_df.groupby(["Make", "Model"])
    .size()
    .reset_index(name="Total_Incidents")
    .sort_values(by="Total_Incidents", ascending=False)
)

make_model_df.head(10)

,Make,Model,Total_Incidents
2465,CESSNA,152,2229
2488,CESSNA,172,1649
2539,CESSNA,172N,1093
5264,PIPER,PA-28-140,863
2537,CESSNA,172M,759
2438,CESSNA,150,752
2542,CESSNA,172P,665
2594,CESSNA,182,617
2571,CESSNA,180,595
2464,CESSNA,150M,551


* Because models are not unique to make as shown above, we need a unique identifier for a given plane
I will combine make and model to form a unique identifier, eg `CESSNA 152`

In [26]:
# Create a unique identifier combining Make and Model
filtered_df["Make_Model"] = (
    filtered_df["Make"].astype(str).str.strip()
    + " "
    + filtered_df["Model"].astype(str).str.strip()
)

# Inspect top unique plane types and total distinct counts
print(f"Total unique plane types: {filtered_df['Make_Model'].nunique():,}\n")
print(filtered_df["Make_Model"].value_counts().head(10))

Total unique plane types: 6,424

Make_Model
CESSNA 152         2229
CESSNA 172         1649
CESSNA 172N        1093
PIPER PA-28-140     863
CESSNA 172M         759
CESSNA 150          752
CESSNA 172P         665
CESSNA 182          617
CESSNA 180          595
CESSNA 150M         551
Name: count, dtype: int64


In [27]:
filtered_df.head()

,Event.Id,Event.Date,Aircraft.Category,Amateur.Built,Make,Model,Aircraft.damage,Injury.Severity,Total.Fatal.Injuries,Total.Serious.Injuries,...,Number.of.Engines,Engine.Type,FAR.Description,Weather.Condition,Broad.phase.of.flight,Purpose.of.flight,Year,Total_Occupants,Is_Destroyed,Make_Model
3601,20001214X42095,1983-01-01,NaN,No,CESSNA,182P,SUBSTANTIAL,NON-FATAL,0.0,0.0,...,1.0,Reciprocating,NaN,VMC,Approach,Personal,1983,4.0,0,CESSNA 182P
3602,20001214X42067,1983-01-01,NaN,No,CESSNA,182RG,SUBSTANTIAL,NON-FATAL,0.0,0.0,...,1.0,Reciprocating,NaN,VMC,Landing,Personal,1983,2.0,0,CESSNA 182RG
3603,20001214X42063,1983-01-01,NaN,No,CESSNA,182P,SUBSTANTIAL,NON-FATAL,0.0,0.0,...,1.0,Reciprocating,NaN,VMC,Takeoff,Personal,1983,1.0,0,CESSNA 182P
3604,20001214X42018,1983-01-01,NaN,No,PIPER,PA-28R-200,SUBSTANTIAL,NON-FATAL,0.0,0.0,...,1.0,Reciprocating,NaN,VMC,Approach,Personal,1983,2.0,0,PIPER PA-28R-200
3605,20001214X41951,1983-01-01,NaN,No,CESSNA,140,SUBSTANTIAL,NON-FATAL,0.0,0.0,...,1.0,Reciprocating,NaN,VMC,Landing,Instructional,1983,2.0,0,CESSNA 140


### Cleaning other columns
- there are other columns containing data that might be related to the outcome of an accident. We list a few here:
- Engine.Type
- Weather.Condition
- Number.of.Engines
- Purpose.of.flight
- Broad.phase.of.flight

Inspect and identify potential cleaning tasks in each of the above columns. Execute those cleaning tasks. 

**Note**: You do not necessarily need to impute or drop NaNs here.

#### 1. cleaning Engine Type
* View counts of different values
* Fill nulls
* Standardize values

In [28]:
filtered_df['Engine.Type'].value_counts()

Engine.Type
Reciprocating      55079
Turbo Shaft         2989
Turbo Prop          2796
Turbo Fan           2173
Unknown             1513
Turbo Jet            543
Geared Turbofan       12
LR                     1
UNK                    1
NONE                   1
Name: count, dtype: int64

In [29]:
def clean_column_engine_type(df):
    df["Engine.Type"] = (
        df["Engine.Type"].fillna("UNKNOWN").astype(str).str.strip().str.upper()
    )
    unknowns = df["Engine.Type"].isin(["UNK", "NONE", "LR", ""])
    df.loc[unknowns, "Engine.Type"] = "UNKNOWN"

    return df



filtered_df = clean_column_engine_type(filtered_df)
filtered_df['Engine.Type'].value_counts()  

Engine.Type
RECIPROCATING      55079
UNKNOWN             6953
TURBO SHAFT         2989
TURBO PROP          2796
TURBO FAN           2173
TURBO JET            543
GEARED TURBOFAN       12
Name: count, dtype: int64

#### 2. Cleaning Weather column
* View counts of different values
* Fill nulls
* Standardize values

In [30]:
filtered_df['Weather.Condition'].value_counts()

Weather.Condition
VMC    60637
IMC     5261
UNK      704
Unk      229
Name: count, dtype: int64

In [31]:
# 1. Convert to uppercase (merges 'Unk' into 'UNK')
filtered_df["Weather.Condition"] = (
    filtered_df["Weather.Condition"]
    .fillna("UNKNOWN")
    .astype(str)
    .str.strip()
    .str.upper()
)

# 2. Replace 'UNK' with 'UNKNOWN'
filtered_df["Weather.Condition"] = filtered_df["Weather.Condition"].replace(
    "UNK", "UNKNOWN"
)

# Verify updated counts
print(filtered_df["Weather.Condition"].value_counts())

Weather.Condition
VMC        60637
IMC         5261
UNKNOWN     4647
Name: count, dtype: int64


#### 3. Cleaning Number of Engines column
* View value counts
* Impute NaNs


In [32]:
filtered_df['Number.of.Engines'].value_counts()

Number.of.Engines
1.0    54519
2.0     9611
0.0      741
3.0      431
4.0      401
8.0        1
Name: count, dtype: int64

Since planes need at least one engine to fly , I will convert the 0.0 to NaN then impute with the median which was 1.00 in our `.describe`

In [33]:
#convert 0.0 to NaN
filtered_df['Number.of.Engines'] = filtered_df['Number.of.Engines'].replace(0.0, np.nan)
# 2. Impute NaNs with median 
filtered_df["Number.of.Engines"] = filtered_df["Number.of.Engines"].fillna(filtered_df["Number.of.Engines"].median())
filtered_df['Number.of.Engines'].value_counts()

Number.of.Engines
1.0    60101
2.0     9611
3.0      431
4.0      401
8.0        1
Name: count, dtype: int64

#### 4. Cleaning Purpose of flight column
* Impute nulls
* Standardize categories

In [34]:
filtered_df['Purpose.of.flight'].value_counts()

Purpose.of.flight
Personal                     36954
Instructional                 9423
Unknown                       5506
Aerial Application            4134
Business                      3431
Positioning                   1446
Other Work Use                1061
Aerial Observation             709
Public Aircraft                643
Ferry                          640
Executive/corporate            424
Skydiving                      175
Flight Test                    160
Banner Tow                      97
External Load                   85
Public Aircraft - Federal       75
Public Aircraft - State         57
Public Aircraft - Local         52
Glider Tow                      41
Firefighting                    33
Air Race show                   27
Air Race/show                   18
Air Drop                        10
PUBS                             3
ASHO                             3
PUBL                             1
Name: count, dtype: int64

In [35]:
# 1. Uppercase and strip whitespace
filtered_df["Purpose.of.flight"] = (
    filtered_df["Purpose.of.flight"]
    .fillna("UNKNOWN")
    .astype(str)
    .str.strip()
    .str.upper()
)

# 2. Merge duplicate/public categories
filtered_df["Purpose.of.flight"] = filtered_df["Purpose.of.flight"].replace(
    {
        "AIR RACE SHOW": "AIR RACE / SHOW",
        "AIR RACE/SHOW": "AIR RACE / SHOW",
        "ASHO": "AIR RACE / SHOW",
        "PUBLIC AIRCRAFT - FEDERAL": "PUBLIC AIRCRAFT",
        "PUBLIC AIRCRAFT - STATE": "PUBLIC AIRCRAFT",
        "PUBLIC AIRCRAFT - LOCAL": "PUBLIC AIRCRAFT",
        "PUBS": "PUBLIC AIRCRAFT",
        "PUBL": "PUBLIC AIRCRAFT",
    }
)

# Verify clean counts
print(filtered_df["Purpose.of.flight"].value_counts())

Purpose.of.flight
PERSONAL               36954
UNKNOWN                10843
INSTRUCTIONAL           9423
AERIAL APPLICATION      4134
BUSINESS                3431
POSITIONING             1446
OTHER WORK USE          1061
PUBLIC AIRCRAFT          831
AERIAL OBSERVATION       709
FERRY                    640
EXECUTIVE/CORPORATE      424
SKYDIVING                175
FLIGHT TEST              160
BANNER TOW                97
EXTERNAL LOAD             85
AIR RACE / SHOW           48
GLIDER TOW                41
FIREFIGHTING              33
AIR DROP                  10
Name: count, dtype: int64


#### Cleaning Broad Phase of flight columnn

In [36]:
filtered_df['Broad.phase.of.flight'].value_counts()

Broad.phase.of.flight
Landing        13018
Takeoff         9801
Cruise          8397
Maneuvering     6267
Approach        5207
Taxi            1670
Climb           1656
Descent         1561
Go-around       1185
Standing         830
Unknown          394
Other             79
Name: count, dtype: int64

In [37]:
filtered_df['Broad.phase.of.flight'] = filtered_df['Broad.phase.of.flight'].fillna("UNKNOWN").astype(str).str.strip().str.upper()
filtered_df['Broad.phase.of.flight'] = filtered_df['Broad.phase.of.flight'].replace({"OTHER": "UNKNOWN"})
filtered_df['Broad.phase.of.flight'].value_counts()

Broad.phase.of.flight
UNKNOWN        20953
LANDING        13018
TAKEOFF         9801
CRUISE          8397
MANEUVERING     6267
APPROACH        5207
TAXI            1670
CLIMB           1656
DESCENT         1561
GO-AROUND       1185
STANDING         830
Name: count, dtype: int64

### Column Removal
- inspect the dataframe and drop any columns that have too many NaNs

In [38]:
#view columns and number of NaNs
filtered_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 70545 entries, 3601 to 88888
Data columns (total 22 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   Event.Id                70545 non-null  object        
 1   Event.Date              70545 non-null  datetime64[ns]
 2   Aircraft.Category       21618 non-null  object        
 3   Amateur.Built           70545 non-null  object        
 4   Make                    70545 non-null  object        
 5   Model                   70545 non-null  object        
 6   Aircraft.damage         70545 non-null  object        
 7   Injury.Severity         70545 non-null  object        
 8   Total.Fatal.Injuries    70545 non-null  float64       
 9   Total.Serious.Injuries  70545 non-null  float64       
 10  Total.Minor.Injuries    70545 non-null  float64       
 11  Total.Uninjured         70545 non-null  float64       
 12  Number.of.Engines       70545 non-null  float64 

#### Columns to drop : 
* Amatuer.Built - we already filtered to have only professional builds so the column contains a single constant value ("NO") is redundant
* FAR.Description - too many nulls,
* Aircraft.Category - too many nulls


In [39]:
filtered_df = filtered_df.drop(columns=['Amateur.Built', 'FAR.Description',  'Aircraft.Category'])
filtered_df.columns

Index(['Event.Id', 'Event.Date', 'Make', 'Model', 'Aircraft.damage',
       'Injury.Severity', 'Total.Fatal.Injuries', 'Total.Serious.Injuries',
       'Total.Minor.Injuries', 'Total.Uninjured', 'Number.of.Engines',
       'Engine.Type', 'Weather.Condition', 'Broad.phase.of.flight',
       'Purpose.of.flight', 'Year', 'Total_Occupants', 'Is_Destroyed',
       'Make_Model'],
      dtype='object')

### Save DataFrame to csv
- its generally useful to save data to file/server after its in a sufficiently cleaned or intermediate state
- the data can then be loaded directly in another notebook for further analysis
- this helps keep your notebooks and workflow readable, clean and modularized

In [40]:
#save clean data to csv
filtered_df.to_csv("cleaned_aviation_data.csv", index=False)
